lib/ocpp/v16/charge_point.cpp

```diff
@@ -2162,6 +2320,435 @@ std::map<int32_t, ChargingSchedule> ChargePoint::get_all_composite_charging_sche
+
+    ocpp::v201::AuthorizeRequest authorize_req;
+    ocpp::v201::IdToken id_token;
+
+    id_token.type = ocpp::v201::IdTokenEnum::eMAID;
+    id_token.idToken = emaid;
+    authorize_req.iso15118CertificateHashData = iso15118_certificate_hash_data;
+    authorize_req.idToken = id_token;
+
+    // do this if C07.FR.06
+    // If Charging Station is not able to validate a contract certificate, because it does not have the associated root
+    // certificate AND CentralContractValidationAllowed is true
+
+    if (this->configuration->getCentralContractValidationAllowed().has_value() and
+        this->configuration->getCentralContractValidationAllowed().value()) {
+        authorize_req.certificate = certificate;
+    }
+
+    req.data.emplace(json(authorize_req).dump());
+
+    Call<DataTransferRequest> call(req, this->message_queue->createMessageId());
+    auto authorize_future = this->send_async<DataTransferRequest>(call);
+
+    auto enhanced_message = authorize_future.get();
+
+    if (enhanced_message.messageType == MessageType::DataTransferResponse) {
+        try {
+            // parse and return authorize response
+            ocpp::CallResult<DataTransferResponse> call_result = enhanced_message.message;
+            if (call_result.msg.data.has_value()) {
+                authorize_response = json::parse(call_result.msg.data.value());
+                return authorize_response;
+            } else {
+                EVLOG_warning << "CSMS response of DataTransferRequest(Authorize) did not include data";
+            }
+        } catch (const json::exception& e) {
+            EVLOG_warning << "Could not parse data of DataTransfer message Authorize.conf: " << e.what();
+        } catch (const std::exception& e) {
+            EVLOG_error << "Unknown Error while handling DataTransfer message Authorize.conf: " << e.what();
+        }
+
+    } else if (enhanced_message.offline) {
+        if (this->configuration->getContractValidationOffline()) {
+            // C07.FR.08
+            // The Charging Station SHALL try to validate the contract certificate locally.
+
+            // if valid: C07.FR.09
+            if (this->configuration->getLocalAuthorizeOffline()) {
+                // use auth cache to look up emaid
+            }
+        } else {
+            // C07.FR.07
+            // The Charging Station SHALL NOT allow charging.
+        }
+    } else {
+        EVLOG_warning << "CSMS response of DataTransferRequest(Authorize) was not DataTransferResponse";
+    }
+    return authorize_response; // FIXME(piet)
+}
+
+void ChargePoint::data_transfer_pnc_sign_certificate() {
+
+    DataTransferRequest req;
+    req.vendorId = ISO15118_PNC_VENDOR_ID;
+    req.messageId.emplace(CiString<50>(std::string("SignCertificate")));
+
+    ocpp::v201::SignCertificateRequest csr_req;
+
+    // TODO(piet): Fix csr data
+    const auto csr = this->pki_handler->generateCsr(ocpp::CertificateSigningUseEnum::V2GCertificate, "DE", "BW",
```

Comment:
Should we really hardcode these values for the moment?

Reply:
This will change anyways with the Hubject extension for additional configuration keys. I would keep this as it is now and change that with the PR for that.